src/server/zset_family.cc

```diff
@@ -870,61 +898,187 @@ OpResult<AddResult> OpAdd(const OpArgs& op_args, const ZParams& zparams, string_
+  }
+
+  for (unsigned i = 0; i < op_args->num_keys; ++i) {
+    string_view weight = ArgS(args, i + 1);
+    if (!absl::SimpleAtod(weight, &(op_args->weights[i]))) {
+      return OpStatus::INVALID_FLOAT;
+    }
+  }
+  return op_args->num_keys;
+}
+
+OpResult<unsigned> ValidateKeyCount(string_view arg_num_keys) {
   // we parsed the structure before, when transaction has been initialized.
-  CHECK(absl::SimpleAtoi(num_str, &store_args.num_keys));
-  DCHECK_GE(args.size(), 3 + store_args.num_keys);
+  unsigned num_keys;
+  if (!absl::SimpleAtoi(arg_num_keys, &num_keys)) {
+    return OpStatus::SYNTAX_ERR;
+  }
+  return num_keys;
+}
+
+OpResult<unsigned> ParseWithScores(CmdArgList args, SetOpArgs* op_args) {
+  if (args.size() != 1) {
+    return OpStatus::SYNTAX_ERR;
+  }
+  op_args->with_scores = true;
+  return 0;
+}
+
+OpResult<SetOpArgs> ParseSetOpArgs(CmdArgList args, bool store) {
+  // TODO: support variadic key for ZUNION command (now fixed to 3)
+  string_view num_str = store ? ArgS(args, 2) : "3";
+  SetOpArgs op_args;
 
-  store_args.weights.resize(store_args.num_keys, 1);
-  for (size_t i = 3 + store_args.num_keys; i < args.size(); ++i) {
+  const auto parsed = ValidateKeyCount(num_str);
+  if (!parsed) {
+    return parsed.status();
+  }
+  op_args.num_keys = *parsed;
+  unsigned key_start = store ? 3 + op_args.num_keys : 1 + op_args.num_keys;
+  // TODO: modify this check when there is variadic key support for ZUNION command
+  DCHECK_GE(args.size(), key_start);
+
+  op_args.weights.resize(op_args.num_keys, 1);
+
+  for (size_t i = key_start; i < args.size(); ++i) {
     ToUpper(&args[i]);
     string_view arg = ArgS(args, i);
     if (arg == "WEIGHTS") {
-      if (args.size() <= i + store_args.num_keys) {
-        return OpStatus::SYNTAX_ERR;
-      }
-
-      for (unsigned j = 0; j < store_args.num_keys; ++j) {
-        string_view weight = ArgS(args, i + j + 1);
-        if (!absl::SimpleAtod(weight, &store_args.weights[j])) {
-          return OpStatus::INVALID_FLOAT;
-        }
+      const auto parsed = ParseWeights(args.subspan(i), &op_args);
+      if (!parsed) {
+        return parsed.status();
       }
-      i += store_args.num_keys;
+      i += *parsed;
     } else if (arg == "AGGREGATE") {
-      if (i + 2 != args.size()) {
-        return OpStatus::SYNTAX_ERR;
+      const auto parsed = ParseAggregate(args.subspan(i), store, &op_args);
+      if (!parsed) {
+        return parsed.status();
       }
-
-      ToUpper(&args[i + 1]);
-
-      string_view agg = ArgS(args, i + 1);
-      if (agg == "SUM") {
-        store_args.agg_type = AggType::SUM;
-      } else if (agg == "MIN") {
-        store_args.agg_type = AggType::MIN;
-      } else if (agg == "MAX") {
-        store_args.agg_type = AggType::MAX;
-      } else {
-        return OpStatus::SYNTAX_ERR;
+      i += *parsed;
+      // Commands with store capability does not offer WITHSCORES option
+      if (store) {
+        break;
       }
```

Comment:
I believe this break can be removed. If store is true and WITHSCORE arg is send you will return syntax error